src/foam/mlang/mlang.js

```diff
@@ -3240,10 +3277,40 @@ foam.CLASS({
     {
       name: 'f',
       code: function(o) {
+        if ( foam.core.Reference.isInstance(this.arg1) ) {
+          return o[property.name + '$find'].then(val => this.arg2.f(val));
+        }
         return this.arg2.f(this.arg1.f(o));
       },
       javaCode: `
-        return getArg2().f(getArg1().f(obj));
+        StringBuilder sb = new StringBuilder("find");
+        PropertyInfo p1 = (PropertyInfo) getArg1();
+        FObject obj1;
+        if ( p1 instanceof AbstractFObjectPropertyInfo ) {
+          Object val = getArg1().f(obj);
+          if ( val == null )
```

Comment:
A bit nicer: return val == null ? : getArg2().f(val);

But your code, so you can choose.